File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseWriter.java
package com.apollographql.apollo.internal.response;

import com.apollographql.apollo.response.CustomTypeAdapter;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.ResponseField;
import com.apollographql.apollo.api.ResponseFieldMarshaller;
import com.apollographql.apollo.api.ResponseWriter;
import com.apollographql.apollo.api.ScalarType;
import com.apollographql.apollo.api.internal.Optional;
import com.apollographql.apollo.response.ScalarTypeAdapters;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class RealResponseWriter implements ResponseWriter {
  private final Operation.Variables operationVariables;
  private final ScalarTypeAdapters scalarTypeAdapters;
  final Map<String, FieldDescriptor> buffer = new LinkedHashMap<>();

  public RealResponseWriter(Operation.Variables operationVariables, ScalarTypeAdapters scalarTypeAdapters) {
    this.operationVariables = operationVariables;
    this.scalarTypeAdapters = scalarTypeAdapters;
  }

  @Override public void writeString(@NotNull ResponseField field, @Nullable String value) {
    writeScalarFieldValue(field, value);
  }

  @Override public void writeInt(@NotNull ResponseField field, @Nullable Integer value) {
    writeScalarFieldValue(field, value != null ? BigDecimal.valueOf(value) : null);
  }

  @Override public void writeLong(@NotNull ResponseField field, @Nullable Long value) {
    writeScalarFieldValue(field, value != null ? BigDecimal.valueOf(value) : null);
  }

  @Override public void writeDouble(@NotNull ResponseField field, @Nullable Double value) {
    writeScalarFieldValue(field, value != null ? BigDecimal.valueOf(value) : null);
  }

  @Override public void writeBoolean(@NotNull ResponseField field, @Nullable Boolean value) {
    writeScalarFieldValue(field, value);
  }

  @SuppressWarnings("unchecked")
  @Override public void writeCustom(@NotNull ResponseField.CustomTypeField field, @Nullable Object value) {
    CustomTypeAdapter typeAdapter = scalarTypeAdapters.adapterFor(field.scalarType());
    writeScalarFieldValue(field, value != null ? typeAdapter.encode(value).value : null);
  }

  @Override public void writeObject(@NotNull ResponseField field, @Nullable ResponseFieldMarshaller marshaller) {
    checkFieldValue(field, marshaller);
    if (marshaller == null) {
      buffer.put(field.responseName(), new FieldDescriptor(field, null));
      return;
    }


    RealResponseWriter nestedResponseWriter = new RealResponseWriter(operationVariables, scalarTypeAdapters);
    marshaller.marshal(nestedResponseWriter);

    buffer.put(field.responseName(), new FieldDescriptor(field, nestedResponseWriter.buffer));
  }

  @Override
  public void writeList(@NotNull ResponseField field, @Nullable List values, @NotNull ListWriter listWriter) {
    checkFieldValue(field, values);

    if (values == null) {
      buffer.put(field.responseName(), new FieldDescriptor(field, null));
      return;
    }

    List accumulated = new ArrayList();
    listWriter.write(values, new ListItemWriter(operationVariables, scalarTypeAdapters, accumulated));
    buffer.put(field.responseName(), new FieldDescriptor(field, accumulated));
  }

  public void resolveFields(ResolveDelegate<Map<String, Object>> delegate) {
    resolveFields(operationVariables, delegate, buffer);
  }

  private void writeScalarFieldValue(ResponseField field, Object value) {
    checkFieldValue(field, value);
    buffer.put(field.responseName(), new FieldDescriptor(field, value));
  }

  @SuppressWarnings("unchecked")
  private Map<String, Object> rawFieldValues(Map<String, FieldDescriptor> buffer) {
    Map<String, Object> fieldValues = new LinkedHashMap<>();
    for (Map.Entry<String, FieldDescriptor> entry : buffer.entrySet()) {
      String fieldResponseName = entry.getKey();
      Object fieldValue = entry.getValue().value;
      if (fieldValue == null) {
        fieldValues.put(fieldResponseName, null);
      } else if (fieldValue instanceof Map) {
        Map<String, Object> nestedMap = rawFieldValues((Map<String, FieldDescriptor>) fieldValue);
        fieldValues.put(fieldResponseName, nestedMap);
      } else if (fieldValue instanceof List) {
        fieldValues.put(fieldResponseName, rawListFieldValues((List) fieldValue));
      } else {
        fieldValues.put(fieldResponseName, fieldValue);
      }
    }
    return fieldValues;
  }

  @SuppressWarnings("unchecked") private List rawListFieldValues(List values) {
    List listValues = new ArrayList();
    for (Object value : values) {
      if (value instanceof Map) {
        listValues.add(rawFieldValues((Map<String, FieldDescriptor>) value));
      } else if (value instanceof List) {
        listValues.add(rawListFieldValues((List) value));
      } else {
        listValues.add(value);
      }
    }
    return listValues;
  }

  @SuppressWarnings("unchecked") private void resolveFields(Operation.Variables operationVariables,
      ResolveDelegate<Map<String, Object>> delegate, Map<String, FieldDescriptor> buffer) {
    Map<String, Object> rawFieldValues = rawFieldValues(buffer);
    for (String fieldResponseName : buffer.keySet()) {
      FieldDescriptor fieldDescriptor = buffer.get(fieldResponseName);
      Object rawFieldValue = rawFieldValues.get(fieldResponseName);
      delegate.willResolve(fieldDescriptor.field, operationVariables);

      switch (fieldDescriptor.field.type()) {
        case OBJECT: {
          resolveObjectFields(fieldDescriptor, (Map<String, Object>) rawFieldValue, delegate);
          break;
        }

        case LIST: {
          resolveListField(fieldDescriptor.field, (List) fieldDescriptor.value, (List) rawFieldValue, delegate);
          break;
        }

        default: {
          if (rawFieldValue == null) {
            delegate.didResolveNull();
          } else {
            delegate.didResolveScalar(rawFieldValue);
          }
          break;
        }
      }

      delegate.didResolve(fieldDescriptor.field, operationVariables);
    }
  }

  @SuppressWarnings("unchecked") private void resolveObjectFields(FieldDescriptor fieldDescriptor,
      Map<String, Object> rawFieldValues, ResolveDelegate<Map<String, Object>> delegate) {
    delegate.willResolveObject(fieldDescriptor.field, Optional.fromNullable(rawFieldValues));
    if (fieldDescriptor.value == null) {
      delegate.didResolveNull();
    } else {
      resolveFields(operationVariables, delegate, (Map<String, FieldDescriptor>) fieldDescriptor.value);
    }
    delegate.didResolveObject(fieldDescriptor.field, Optional.fromNullable(rawFieldValues));
  }

  @SuppressWarnings("unchecked") private void resolveListField(ResponseField listResponseField, List fieldValues,
      List rawFieldValues, ResolveDelegate<Map<String, Object>> delegate) {
    if (fieldValues == null) {
      delegate.didResolveNull();
      return;
    }

    for (int i = 0; i < fieldValues.size(); i++) {
      delegate.willResolveElement(i);

      Object fieldValue = fieldValues.get(i);
      if (fieldValue instanceof Map) {
        delegate.willResolveObject(listResponseField,
            Optional.fromNullable((Map<String, Object>) rawFieldValues.get(i)));
        resolveFields(operationVariables, delegate, (Map<String, FieldDescriptor>) fieldValue);
        delegate.didResolveObject(listResponseField,
            Optional.fromNullable((Map<String, Object>) rawFieldValues.get(i)));
      } else if (fieldValue instanceof List) {
        resolveListField(listResponseField, (List) fieldValue, (List) rawFieldValues.get(i), delegate);
      } else {
        delegate.didResolveScalar(rawFieldValues.get(i));
      }
      delegate.didResolveElement(i);
    }
    delegate.didResolveList(rawFieldValues);
  }

  private static void checkFieldValue(ResponseField field, Object value) {
    if (!field.optional() && value == null) {
      throw new NullPointerException(String.format("Mandatory response field `%s` resolved with null value",
          field.responseName()));
    }
  }

  @SuppressWarnings("unchecked") private static final class ListItemWriter implements ResponseWriter.ListItemWriter {
    final Operation.Variables operationVariables;
    final com.apollographql.apollo.response.ScalarTypeAdapters scalarTypeAdapters;
    final List accumulator;

    ListItemWriter(Operation.Variables operationVariables, ScalarTypeAdapters scalarTypeAdapters, List accumulator) {
      this.operationVariables = operationVariables;
      this.scalarTypeAdapters = scalarTypeAdapters;
      this.accumulator = accumulator;
    }

    @Override public void writeString(@Nullable String value) {
      accumulator.add(value);
    }

    @Override public void writeInt(@Nullable Integer value) {
      accumulator.add(value != null ? BigDecimal.valueOf(value) : null);
    }

    @Override public void writeLong(@Nullable Long value) {
      accumulator.add(value != null ? BigDecimal.valueOf(value) : null);
    }

    @Override public void writeDouble(@Nullable Double value) {
      accumulator.add(value != null ? BigDecimal.valueOf(value) : null);
    }

    @Override public void writeBoolean(@Nullable Boolean value) {
      accumulator.add(value);
    }

    @Override public void writeCustom(@NotNull ScalarType scalarType, @Nullable Object value) {
      CustomTypeAdapter typeAdapter = scalarTypeAdapters.adapterFor(scalarType);
      accumulator.add(value != null ? typeAdapter.encode(value).value : null);
    }

    @Override public void writeObject(ResponseFieldMarshaller marshaller) {
      RealResponseWriter nestedResponseWriter = new RealResponseWriter(operationVariables, scalarTypeAdapters);
      marshaller.marshal(nestedResponseWriter);
      accumulator.add(nestedResponseWriter.buffer);
    }

    @Override public void writeList(@Nullable List items, @NotNull ResponseWriter.ListWriter listWriter) {
      if (items == null) {
        accumulator.add(null);
      } else {
        List nestedAccumulated = new ArrayList();
        listWriter.write(items, new ListItemWriter(operationVariables, scalarTypeAdapters, nestedAccumulated));
        accumulator.add(nestedAccumulated);
      }
    }
  }

  private static class FieldDescriptor {
    final ResponseField field;
    final Object value;

    FieldDescriptor(ResponseField field, Object value) {
      this.field = field;
      this.value = value;
    }
  }
}

File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseWriter.kt
package com.apollographql.apollo.internal.response

import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.ScalarType
import com.apollographql.apollo.api.ScalarTypeAdapters
import com.apollographql.apollo.api.internal.ResolveDelegate
import com.apollographql.apollo.api.internal.ResponseFieldMarshaller
import com.apollographql.apollo.api.internal.ResponseWriter
import java.math.BigDecimal
import java.util.ArrayList
import java.util.LinkedHashMap

class RealResponseWriter(private val operationVariables: Operation.Variables, private val scalarTypeAdapters: ScalarTypeAdapters) : ResponseWriter {
  val buffer: MutableMap<String, FieldDescriptor> = LinkedHashMap()
  override fun writeString(field: ResponseField, value: String?) {
    writeScalarFieldValue(field, value)
  }

  override fun writeInt(field: ResponseField, value: Int?) {
    writeScalarFieldValue(field, if (value != null) BigDecimal.valueOf(value.toLong()) else null)
  }

  override fun writeLong(field: ResponseField, value: Long?) {
    writeScalarFieldValue(field, if (value != null) BigDecimal.valueOf(value) else null)
  }

  override fun writeDouble(field: ResponseField, value: Double?) {
    writeScalarFieldValue(field, if (value != null) BigDecimal.valueOf(value) else null)
  }

  override fun writeBoolean(field: ResponseField, value: Boolean?) {
    writeScalarFieldValue(field, value)
  }

  override fun writeCustom(field: ResponseField.CustomTypeField, value: Any?) {
    val typeAdapter = scalarTypeAdapters.adapterFor<Any>(field.scalarType)
    writeScalarFieldValue(field, if (value != null) typeAdapter.encode(value).value else null)
  }

  override fun writeObject(field: ResponseField, marshaller: ResponseFieldMarshaller?) {
    checkFieldValue(field, marshaller)
    if (marshaller == null) {
      buffer[field.responseName] = FieldDescriptor(field, null)
      return
    }
    val nestedResponseWriter = RealResponseWriter(operationVariables, scalarTypeAdapters)
    marshaller.marshal(nestedResponseWriter)
    buffer[field.responseName] = deepMergeObjects(field, buffer[field.responseName]?.value, nestedResponseWriter.buffer)
  }

  private fun deepMergeObjects(field: ResponseField, oldValue: Any?, newValue: Map<String, FieldDescriptor>): FieldDescriptor {
    return if (oldValue == null || oldValue !is Map<*, *>) {
      FieldDescriptor(field, newValue)
    } else {
      val oldMap = oldValue as Map<String, FieldDescriptor>

      val mergedCommonValues = oldMap.keys.intersect(newValue.keys)
          .filter { newValue[it]?.value is Map<*, *> }
          .map { deepMergeObjects(oldMap[it]!!.field, oldMap[it]?.value, newValue[it]!!.value as Map<String, FieldDescriptor>) }
          .associateBy { it.field.responseName }

      FieldDescriptor(field, oldMap + newValue + mergedCommonValues)
    }
  }

  override fun writeFragment(marshaller: ResponseFieldMarshaller?) {
    marshaller?.marshal(this)
  }

  override fun <T> writeList(field: ResponseField, values: List<T>?, listWriter: ResponseWriter.ListWriter<T>) {
    checkFieldValue(field, values)
    if (values == null) {
      buffer[field.responseName] = FieldDescriptor(field, null)
      return
    }
    val accumulated = ArrayList<Any?>()
    listWriter.write(values, ListItemWriter(operationVariables, scalarTypeAdapters, accumulated))
    buffer[field.responseName] = FieldDescriptor(field, accumulated)
  }

  fun resolveFields(delegate: ResolveDelegate<Map<String, Any>?>) {
    resolveFields(operationVariables, delegate, buffer)
  }

  private fun writeScalarFieldValue(field: ResponseField, value: Any?) {
    checkFieldValue(field, value)
    buffer[field.responseName] = FieldDescriptor(field, value)
  }

  private fun rawFieldValues(buffer: Map<String, FieldDescriptor>): Map<String, Any?> {
    val fieldValues: MutableMap<String, Any?> = LinkedHashMap()
    for ((fieldResponseName, value) in buffer) {
      val fieldValue = value.value
      if (fieldValue == null) {
        fieldValues[fieldResponseName] = null
      } else if (fieldValue is Map<*, *>) {
        val nestedMap = rawFieldValues(fieldValue as Map<String, FieldDescriptor>)
        fieldValues[fieldResponseName] = nestedMap
      } else if (fieldValue is List<*>) {
        fieldValues[fieldResponseName] = rawListFieldValues(fieldValue)
      } else {
        fieldValues[fieldResponseName] = fieldValue
      }
    }
    return fieldValues
  }

  private fun rawListFieldValues(values: List<*>): List<*> {
    val listValues = ArrayList<Any?>()
    for (value in values) {
      if (value is Map<*, *>) {
        listValues.add(rawFieldValues(value as Map<String, FieldDescriptor>))
      } else if (value is List<*>) {
        listValues.add(rawListFieldValues(value))
      } else {
        listValues.add(value)
      }
    }
    return listValues
  }

  private fun resolveFields(operationVariables: Operation.Variables,
                            delegate: ResolveDelegate<Map<String, Any>?>, buffer: Map<String, FieldDescriptor>) {
    val rawFieldValues = rawFieldValues(buffer)
    for (fieldResponseName in buffer.keys) {
      val fieldDescriptor = buffer[fieldResponseName]
      val rawFieldValue = rawFieldValues[fieldResponseName]
      delegate.willResolve(fieldDescriptor!!.field, operationVariables, fieldDescriptor.value)
      when (fieldDescriptor.field.type) {
        ResponseField.Type.OBJECT -> {
          resolveObjectFields(fieldDescriptor, rawFieldValue as Map<String, Any>?, delegate)
        }
        ResponseField.Type.LIST -> {
          resolveListField(fieldDescriptor.field, fieldDescriptor.value as List<*>?, rawFieldValue as List<*>?, delegate)
        }
        else -> {
          if (rawFieldValue == null) {
            delegate.didResolveNull()
          } else {
            delegate.didResolveScalar(rawFieldValue)
          }
        }
      }
      delegate.didResolve(fieldDescriptor.field, operationVariables)
    }
  }

  private fun resolveObjectFields(fieldDescriptor: FieldDescriptor,
                                  rawFieldValues: Map<String, Any>?, delegate: ResolveDelegate<Map<String, Any>?>) {
    delegate.willResolveObject(fieldDescriptor.field, rawFieldValues)
    val value = fieldDescriptor.value
    if (value == null) {
      delegate.didResolveNull()
    } else {
      resolveFields(operationVariables, delegate, value as Map<String, FieldDescriptor>)
    }
    delegate.didResolveObject(fieldDescriptor.field, rawFieldValues)
  }

  private fun resolveListField(listResponseField: ResponseField, fieldValues: List<*>?,
                               rawFieldValues: List<*>?, delegate: ResolveDelegate<Map<String, Any>?>) {
    if (fieldValues == null) {
      delegate.didResolveNull()
      return
    }
    fieldValues.forEachIndexed { i, fieldValue ->
      delegate.willResolveElement(i)

      when (fieldValue) {
        is Map<*, *> -> {
          delegate.willResolveObject(listResponseField, rawFieldValues!![i] as Map<String, Any>?)
          resolveFields(operationVariables, delegate, fieldValue as Map<String, FieldDescriptor>)
          delegate.didResolveObject(listResponseField, rawFieldValues[i] as Map<String, Any>?)
        }
        is List<*> -> {
          resolveListField(listResponseField, fieldValue, rawFieldValues!![i] as List<*>?, delegate)
        }
        else -> {
          delegate.didResolveScalar(rawFieldValues!![i])
        }
      }
      delegate.didResolveElement(i)
    }
    delegate.didResolveList(rawFieldValues!!)
  }

  private class ListItemWriter internal constructor(val operationVariables: Operation.Variables, val scalarTypeAdapters: ScalarTypeAdapters, val accumulator: MutableList<Any?>) : ResponseWriter.ListItemWriter {
    override fun writeString(value: String?) {
      accumulator.add(value)
    }

    override fun writeInt(value: Int?) {
      accumulator.add(if (value != null) BigDecimal.valueOf(value.toLong()) else null)
    }

    override fun writeLong(value: Long?) {
      accumulator.add(if (value != null) BigDecimal.valueOf(value) else null)
    }

    override fun writeDouble(value: Double?) {
      accumulator.add(if (value != null) BigDecimal.valueOf(value) else null)
    }

    override fun writeBoolean(value: Boolean?) {
      accumulator.add(value)
    }

    override fun writeCustom(scalarType: ScalarType, value: Any?) {
      val typeAdapter = scalarTypeAdapters.adapterFor<Any>(scalarType)
      accumulator.add(if (value != null) typeAdapter.encode(value).value else null)
    }

    override fun writeObject(marshaller: ResponseFieldMarshaller?) {
      val nestedResponseWriter = RealResponseWriter(operationVariables, scalarTypeAdapters)
      marshaller!!.marshal(nestedResponseWriter)
      accumulator.add(nestedResponseWriter.buffer)
    }

    override fun <T> writeList(items: List<T>?, listWriter: ResponseWriter.ListWriter<T>) {
      if (items == null) {
        accumulator.add(null)
      } else {
        val nestedAccumulated = ArrayList<Any?>()
        listWriter.write(items, ListItemWriter(operationVariables, scalarTypeAdapters, nestedAccumulated))
        accumulator.add(nestedAccumulated)
      }
    }

  }

  class FieldDescriptor internal constructor(val field: ResponseField, val value: Any?)

  companion object {
    private fun checkFieldValue(field: ResponseField, value: Any?) {
      if (!field.optional && value == null) {
        throw NullPointerException(String.format("Mandatory response field `%s` resolved with null value",
            field.responseName))
      }
    }
  }

}
